### PR TITLE
*: bazel-ify cockroachdb

### PR DESCRIPTION
**Pull request by irfansharif**
This commit introduces a new build system to cockroachdb: Bazel.
The hope here is to eventually replace our current Makefile. It sets the
foundation for introducing Bazel in our CI systems for parallelized
builds/test runs, artifact caching, remote builds, and many other such
profits (literally, it could reduce infra spend).

We originally prototyped a pseudo bazel-ified cockroachdb in #52824,
which resulted in a week long effort (with otan@, jamesl@, and myself) to
wrap it up in earnest in #55258. This is the polished version of the
resulting work, the write up for which can be found at
http://go.crdb.dev/bazel-hackweek.

---

Most of the BUILD files here were auto-generated through Gazelle (see
doc). Some of these auto-generated files required further manual
tweaking to get just right (see doc). These were mostly around the packages that
link into c-dependencies, but also the few packages that rely on
auto-generated code (again, see doc). These specific files are the ones
intended for review (they're also the ones responsible for
auto-generation), among a few others. I've listed them out here:

```
	new file:   .bazelrc
	modified:   .gitattributes
	modified:   .gitignore
	new file:   BUILD.bazel
	new file:   DEPS.bzl
	modified:   Makefile
	new file:   WORKSPACE
	new file:   pkg/ccl/gssapiccl/BUILD.bazel
	new file:   pkg/ccl/storageccl/engineccl/BUILD.bazel
	new file:   pkg/cli/BUILD.bazel
	new file:   pkg/cmd/cockroach-oss/BUILD.bazel
	new file:   pkg/cmd/cockroach-short/BUILD.bazel
	new file:   pkg/cmd/cockroach/BUILD.bazel
	new file:   pkg/col/coldata/BUILD.bazel
	new file:   pkg/geo/geoproj/BUILD.bazel
	modified:   pkg/geo/geoproj/geoproj.go
	modified:   pkg/geo/geoproj/proj.cc
	new file:   pkg/geo/geos/BUILD.bazel
	new file:   pkg/sql/colconv/BUILD.bazel
	new file:   pkg/sql/colexec/BUILD.bazel
	new file:   pkg/sql/colexec/COLEXEC.bzl
	modified:   pkg/sql/colexec/and_or_projection_tmpl.go
	modified:   pkg/sql/colexec/any_not_null_agg_tmpl.go
	modified:   pkg/sql/colexec/avg_agg_tmpl.go
	modified:   pkg/sql/colexec/bool_and_or_agg_tmpl.go
	modified:   pkg/sql/colexec/cast_tmpl.go
	new file:   pkg/sql/colexec/colbuilder/BUILD.bazel
	new file:   pkg/sql/colexec/colexec.go
	modified:   pkg/sql/colexec/concat_agg_tmpl.go
	modified:   pkg/sql/colexec/const_tmpl.go
	modified:   pkg/sql/colexec/count_agg_tmpl.go
	modified:   pkg/sql/colexec/default_agg_tmpl.go
	modified:   pkg/sql/colexec/default_cmp_expr_tmpl.go
	modified:   pkg/sql/colexec/default_cmp_proj_ops_tmpl.go
	modified:   pkg/sql/colexec/default_cmp_sel_ops_tmpl.go
	modified:   pkg/sql/colexec/distinct_tmpl.go
	modified:   pkg/sql/colexec/hash_aggregator_tmpl.go
	modified:   pkg/sql/colexec/hash_utils_tmpl.go
	modified:   pkg/sql/colexec/hashjoiner_tmpl.go
	modified:   pkg/sql/colexec/hashtable_tmpl.go
	modified:   pkg/sql/colexec/is_null_ops_tmpl.go
	modified:   pkg/sql/colexec/mergejoinbase_tmpl.go
	modified:   pkg/sql/colexec/mergejoiner_tmpl.go
	modified:   pkg/sql/colexec/min_max_agg_tmpl.go
	modified:   pkg/sql/colexec/ordered_synchronizer_tmpl.go
	modified:   pkg/sql/colexec/proj_const_ops_tmpl.go
	modified:   pkg/sql/colexec/proj_non_const_ops_tmpl.go
	modified:   pkg/sql/colexec/quicksort_tmpl.go
	modified:   pkg/sql/colexec/rank_tmpl.go
	modified:   pkg/sql/colexec/relative_rank_tmpl.go
	modified:   pkg/sql/colexec/row_number_tmpl.go
	modified:   pkg/sql/colexec/rowstovec_tmpl.go
	modified:   pkg/sql/colexec/select_in_tmpl.go
	modified:   pkg/sql/colexec/selection_ops_tmpl.go
	modified:   pkg/sql/colexec/sort_tmpl.go
	modified:   pkg/sql/colexec/substring_tmpl.go
	modified:   pkg/sql/colexec/sum_agg_tmpl.go
	modified:   pkg/sql/colexec/utils_tmpl.go
	modified:   pkg/sql/colexec/values_differ_tmpl.go
	modified:   pkg/sql/colexec/vec_comparators_tmpl.go
	modified:   pkg/sql/colexec/window_peer_grouper_tmpl.go
	new file:   pkg/sql/colflow/BUILD.bazel
	new file:   pkg/sql/opt/BUILD.bazel
	new file:   pkg/sql/opt/exec/BUILD.bazel
	new file:   pkg/sql/opt/exec/execbuilder/BUILD.bazel
	new file:   pkg/sql/opt/exec/explain/BUILD.bazel
	new file:   pkg/sql/opt/memo/BUILD.bazel
	new file:   pkg/sql/opt/norm/BUILD.bazel
	new file:   pkg/sql/opt/optgen/lang/BUILD.bazel
	new file:   pkg/sql/opt/xform/BUILD.bazel
	new file:   pkg/sql/parser/BUILD.bazel
	new file:   pkg/sql/parser/sql-gen.sh
	modified:   pkg/sql/parser/sql.y
	new file:   pkg/storage/BUILD.bazel
	modified:   vendor
```

The changes to all the _tmpl.go files were needed to include necessary
import analysis/compilation for bazel when generating the .eg.go files
(also summarized in doc up above), and should be uncontroversial.

I think it's easiest to pull the branch down locally and to click around
to see what works, and what doesn't. I would start at the top-level
BUILD.bazel and WORSPACE file. I've added comments to explain what most
of added machinery does.

Release note: None

### CODE REVIEW

**Review by petermattis — COMMENTED** [review]
Very exciting!

What is the expectation for how engineers will use `bazel`? Will we change the various `make` targets to using `bazel` rather than `go`? Or will engineers run `bazel` directly?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55687#-:-MK0fieJ072gjOgSYJqY:b4r362b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @jordanlewis, and @otan)

**Comment by irfansharif** [comment]
One thing I didn't mention above is that the current plan here is to have these two separate build systems sitting side-by-side in our codebase for a while. With this PR we think we've done 90% of the work, we only have the remaining 90% left. We'll only know what that is once folks start trying it out in earnest (i.e. after merge), and falling back to `make` as needed. There's no overlap between the two systems whatsoever, by design.

> What is the expectation for how engineers will use bazel? Will we change the various make targets to using bazel rather than go? Or will engineers run bazel directly?

In my view I think we should all want to use bazel directly. I don't think it's too bad, and we can write usable shorthands for the most common things. I also think now that Bazel is a bit more teachable to engineers here than our Makefile is/was, so the more exposure everyone has the better.

**Comment by irfansharif** [comment]
For onlookers, the CI failures are due to #55693 and me somehow breaking the windows build, both of which @jlinder will be looking into.

**Review by petermattis — COMMENTED** [review]
> In my view I think we should all want to use bazel directly. I don't think it's too bad, and we can write usable shorthands for the most common things. I also think that Bazel is a bit more teachable to engineers here than our Makefile is/was, so the more exposure everyone has the better.

I agree on using `bazel` directly. We can also add command aliases for common actions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55687#-:-MK1H9KZ5dDZuKSiCW6j:b-4k3f5u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @jordanlewis, and @otan)

**Comment by otan** [comment]
can we break this up, so we can start committing stuff? happy to take over doing that if you so wish.

ideas:
* add WORKSPACE and root BUILD.bazel
* add c-deps BUILD.bazel && ui rule to check these build
* add vendor update for go-libedit
* gazel everything

should stop the rebase hell, and be isolated by themselves?

**Comment by irfansharif** [comment]
> can we break this up, so we can start committing stuff? 

I did try this above: c-dep changes + picking up go-libedit + everything else (it's _all_ supposed to get reviewed anyway, so yolo, and not like we're expecting to check out individual SHAs in between in the future). I was ok with committing this as is, but I'm also happy to have you take over from here and change it to your liking.

**Comment by irfansharif** [comment]
If you're worried about rebasing, it's not an issue. Rebasing + `bazel run //:gazelle` does the right thing. Everything that needs to be pinned through `# keep` is already done above.

**Comment by petermattis** [comment]
> I was ok with committing this as is, but I'm also happy to have you take over from here and change it to your liking.

Since this is creating a parallel build system and not affecting the make-based build system, I'm ok with merging this PR as-is. It won't disrupt anyone's work, and if we find some fatal flaw we can always revert it.

**Comment by otan** [comment]
heh, ok. are we waiting for rocksdb to go away?

**Comment by petermattis** [comment]
> heh, ok. are we waiting for rocksdb to go away?

🤷 That is in progress as well. I'd expect it to land sometime next week.

**Comment by otan** [comment]
i think this PR can't merge unless we fix rocksdb or wait for it to be deleted. 
i'd rather the latter :D

**Comment by irfansharif** [comment]
Rebased on latest, picking up rocksdb deletion, etc. (#55509, #55858). We also recently updated the builder image to install bazel (#55790). Between the two of these things I'm no longer expecting a windows build failure, nor am I expecting it to trip up our `assert-workspace-clean` linter. Also rebasing on top of #55796 was a bit tedious, so if we could hold off on execgen changes for just a few days I'll buy someone a coffee. Ditto for `vendor` which changes quite often. I'd like to get this in sooner than not.

---

@otan, mind taking another look here? I don't know much about why we needed the changes the fifth commit in particular. I just noticed that in an earlier branch I had sneaked in the following diff
```diff
-// #cgo windows LDFLAGS: -lproj_4_9 -lshlwapi -lrpcrt4
+// #cgo windows LDFLAGS: -lshlwapi -lrpcrt4
```
Probably why I broke windows. Oops.

@jordanlewis, mind signing off on the third commit?

+cc @jlinder, it'd be worthwhile to go through the moving pieces here to pen down details/context for dev-inf going forward.

**Comment by irfansharif** [comment]
Looks like we'll need another change to the builder image: #55910. James is looking to build it and push it out.

File: pkg/sql/colconv/datum_to_vec_tmpl.go
Comment (yuzefovich) [inline pkg/sql/colconv/datum_to_vec_tmpl.go:24]: Some duplicates (and in most files). Also, as mentioned on slack, we probably need to add fake imports like
```
var _ apd.Decimal
```
in order for them to not be formatted away.

Comment (irfansharif) [reply pkg/sql/colconv/datum_to_vec_tmpl.go]: Done.

File: pkg/geo/geoproj/proj.cc
Comment (otan) [inline pkg/geo/geoproj/proj.cc:12]: i think this is actually ok to keep.

**Review by otan — COMMENTED** [review]
if we're okay with BUILD.bazel immediately being stale, this LGTM.

There are some few cosmetic improvements that I'd like to see as a follow up (cc @jlinder) such as:
* Moving c-deps build out of the root BUILD.bazel and into c-deps/BUILD.bazel. Move the configure/cmake stuff out of the root WORKSPACE and into c-deps/BUILD.bazel. Move the third party repositories to a `c-deps/repositories.bzl` rule, similar to how vendor is in there. This keeps everything "self-contained".
* If you wanted to, you should also `s/TODO(irfansharif)/TODO(#bazel)/g` to take yourself out of the game irfan ;)

**Review by otan — APPROVED** [review]

File: pkg/sql/colconv/datum_to_vec_tmpl.go
Comment (yuzefovich) [reply pkg/sql/colconv/datum_to_vec_tmpl.go]: Thanks, that commit LGTM

**Comment by irfansharif** [comment]
> if we're okay with BUILD.bazel immediately being stale, this LGTM.

File: pkg/geo/geoproj/proj.cc
Comment (irfansharif) [reply pkg/geo/geoproj/proj.cc]: Dropped the commit.

**Comment by irfansharif** [comment]
> if we're okay with BUILD.bazel immediately being stale, this LGTM.

It won't be now. Added `bazel run //:gazelle` to `make generate`. We'll actually need this to make the workspace clean thing pass for PRs that change filenames/delete/move files. It should automagically fixup the BUILD files as needed. I'll send out an email to eng@ after this lands. The makefile will warn users if they don't have bazel installed (TODO for @jlinder to pin/assert bazel installation version, 3.7.0 should be fine).

bors r+

**Comment by otan** [comment]
> It won't be now. Added bazel run //:gazelle to make generate.

Please send an email out about this. Probably need to update some docs relating to contributing to CRDB to install `bazel`.

**Comment by otan** [comment]
(namely https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR and the root README.md -- separate PR is ok!)

**Comment by irfansharif** [comment]
Added to https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181338446/Getting+and+building+from+source.

**Comment by irfansharif** [comment]
This was being batched with PRs that added new files. Hm. I'll have to bors it during quiet hours after quickly rebasing off latest, and making sure it's a single PR batch.

**Comment by irfansharif** [comment]
bors single on

File: Makefile
Comment (jin) [inline Makefile:324]: Just passing by! Consider recommending the usage of github.com/bazelbuild/bazelisk and a `.bazelversion` file to ensure your contributors are using the same version.

Great to see CockroachDB using Bazel!

Comment (irfansharif) [reply Makefile:324]: Thanks for the suggestion! Do keep passing on by for more, we're still learning how bazel works so it's all the more welcome, @jin.

**Pull request merged**
